[SYCL][ESIMD] Add test for esimd default constructor

vasilytric opened the pull request:
This PR provides test for `simd` default constructor. 
The test verifies that `simd`'s elements value is equal to default value of element type
Also this PR provides common files:
- function for printing verbose messaged into console
- pre-defined sets of types for testing
- function, that let iterating over pre-defined set of datatypes

vasilytric committed:
[SYCL][ESIMD] Add common code for constructors tests

yuriykoch commented:
Please ensure you follow the [LLVM Developer Policy](https://llvm.org/docs/DeveloperPolicy.html#commit-messages) as required by [CONTRIBUTING.md](https://github.com/intel/llvm-test-suite/blob/intel/CONTRIBUTING.md#commit-message).
More specifically,  aren't the "add common code" and the "add test" commit messages too broad?

Also do we really need a separate "apply clang-format" commit? I mean, it looks like you were preparing the clean commit history before publishing your code anyway..

vasilytric committed:
[SYCL][ESIMD] Add test for default constructor

vasilytric commented:
> Please ensure you follow the [LLVM Developer Policy](https://llvm.org/docs/DeveloperPolicy.html#commit-messages) as required by [CONTRIBUTING.md](https://github.com/intel/llvm-test-suite/blob/intel/CONTRIBUTING.md#commit-message).
> More specifically, aren't the "add common code" and the "add test" commit messages too broad?

Commits messages was changed after force pushing

>Also do we really need a separate "apply clang-format" commit? I mean, it looks like you were preparing the clean commit history before publishing your code anyway..

This commit was created after CI running and clang-format issues was found, but now this commit was deleted due to force pushing

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_default.cpp:
```diff
@@ -0,0 +1,171 @@
+  for_all_types<run_test_with_one_elem>(types, logger, queue, passed);
+  for_all_types<run_test_with_eight_elems>(types, logger, queue, passed);
+  for_all_types<run_test_with_sixteen_elems>(types, logger, queue, passed);
+  for_all_types<run_test_with_thirty_two_elems>(types, logger, queue, passed);
```
```suggestion
  for_all_types<test, 1>(types, logger, queue, passed);
  for_all_types<test, 8>(types, logger, queue, passed);
  for_all_types<test, 16>(types, logger, queue, passed);
  for_all_types<test, 32>(types, logger, queue, passed);
```
Why not something like this?

vasilytric replied:
`for_all_types` can't obtain only type arguments

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_default.cpp:
```diff
@@ -0,0 +1,171 @@
+using run_test_with_thirty_two_elems = test<DataTypeT, 32>;
+
+int main(int argc, char **argv) {
+  logger::Logger logger{};
```
What is the intent behind the logger class instance? Currently the `Logger` class doesn't store any state within.

vasilytric replied:
I've make methods of this class `static` in this commit: 23971e1019c700a93cb02f48c0767d5d1c46c890

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_default.cpp:
```diff
@@ -0,0 +1,171 @@
+template <typename DataTypeT, int TestNumber, int NumElems>
+struct kernel_for_simd;
+
+const auto types =
```
Why isn't this variable in the function scope of the `main`?

vasilytric replied:
I've moved it into separate geter-function in this commit: 4970bd2d73f69c409e7477cccff9a49e4b1da8f8

yuriykoch commented on SYCL/ESIMD/api/functional/logger.hpp:
```diff
@@ -0,0 +1,46 @@
+#include <iostream>
+#include <string>
+
+namespace logger {
```
We already have `esimd_test` namespace. I would suggest to:

1. Use `esimd_test::api::functional` namespace for any header in `SYCL/ESIMD/api/functional`
2. Add  `SYCL/ESIMD/api/functional/common.hpp` to include all general use headers for every functional test

What do you think?

vasilytric replied:
> Add SYCL/ESIMD/api/functional/common.hpp to include all general use headers for every functional test

I'll do that, when I can put something in the `common.hpp` file, I don't want to have an empty ёcommon.hpp` file 

> Use esimd_test::api::functional namespace for any header in SYCL/ESIMD/api/functional


Namespaces was improved in this commit: 19fb0122a5e3efc3edad73f7b1ca0ad3c17c5614

yuriykoch replied:
Thanks)

> I'll do that, when I can put something in the common.hpp file, I don't want to have an empty `common.hpp` file

Actually, there is nothing wrong to have an "include-only" header as well - see `esimd.hpp`

vasilytric replied:
this comment was applied in this commit: 3787c4fc6d62817f7c4e1458c2e5255a60beeef4

yuriykoch commented on SYCL/ESIMD/api/functional/type_coverage.hpp:
```diff
@@ -0,0 +1,49 @@
+          typename... actionArgsT, typename... types, typename... argsT>
+inline void for_all_types(const named_type_pack<types...> &typeList,
+                          argsT &&... args) {
+
+  // run action for each type from types... parameter pack
+  int packExpansion[] = {
+      (action<types, actionArgsT...>{}(std::forward<argsT>(args)...),
+
+       0 // Dummy initialization value
+       )...};
```
```suggestion
  bool succeed{true};
  
  // run action for each type from types... parameter pack
  int packExpansion[] = {
      (succeed &= action<types, actionArgsT...>{}(std::forward<argsT>(args)...),

       0 // Dummy initialization value
       )...};
  // .......
  return succeed;
```

What do you think about something like this? We might incapsulate the `passed` logic into the return values as well.

vasilytric replied:
I've done it in this commit: ebaa8fc5c3830529804a100c915b4e571b57a5d3

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_default.cpp:
```diff
@@ -0,0 +1,171 @@
+
+// Dummy kernel for submitting some code into device side.
+template <typename DataTypeT, int TestNumber, int NumElems>
+struct kernel_for_simd;
```
```suggestion
struct kernel;
```

vasilytric replied:
I've applied this comment in [here](https://github.com/intel/llvm-test-suite/pull/565/commits/35fd849b6fc222cb2da9d90a1dd6185de07651e7#)

yuriykoch replied:
Looks like the lines are still present

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_default.cpp:
```diff
@@ -0,0 +1,171 @@
+      std::vector<DataTypeT, AllocatorT> result{NumElems, AllocatorT{queue}};
+      queue.submit([&](sycl::handler &cgh) {
+        auto out = result.data();
+        cgh.single_task<kernel_for_simd<DataTypeT, 1, NumElems>>([=] {
```
```suggestion
        cgh.single_task<kernel_for_simd<DataTypeT, enum::value, NumElems>>([=] {
```
Magic numbers are going to break some day..

vasilytric replied:
I've applied this comment in this commit: 35fd849b6fc222cb2da9d90a1dd6185de07651e7

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_default.cpp:
```diff
@@ -0,0 +1,171 @@
+          log.fail<DataTypeT, NumElems>(
+              "In simd by initializer elem value is not equal to default "
+              "value, "
+              "it value is: " +
+              std::to_string(it) +
+              ", but provided data type default value is: " +
```
```suggestion
              "retrieved: " +
              std::to_string(it) +
              ", expected: " +
```

As a minor suggestion, here and elsewhere.

vasilytric replied:
I've applied this comment in this commit: 23971e1019c700a93cb02f48c0767d5d1c46c890

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_default.cpp:
```diff
@@ -0,0 +1,171 @@
+      }
+    }
+
+    // verification for initializing simd by rvalue in actual parameter
```
Looks like we have 4 blocks of code with minor differences here. Could we possibly have a templated version of this code, with some "action" and "description" passed?
So that:

- if we want to add 1 more "action" to test, we don't need to copy the entire logic
- if we want to update test logic, we don't need to update it in multiple places

vasilytric replied:
I've applied this comment in this commit: 35fd849b6fc222cb2da9d90a1dd6185de07651e7

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_default.cpp:
```diff
@@ -0,0 +1,171 @@
+//==------- esimd_bit_ops.cpp  - DPC++ ESIMD on-device test ----------------==//
```
Please, update appropriately

vasilytric replied:
I've applied this comment in this commit: 19fb0122a5e3efc3edad73f7b1ca0ad3c17c5614

yuriykoch reviewed (changes requested):
Thanks, looks great in general

yuriykoch commented:
> This commit was created after CI running and clang-format issues was found

My bad. I've missed it, sorry for that

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_default.cpp:
```diff
@@ -0,0 +1,171 @@
+// Use struct with overloaded call operator to have able for iterating over
+// pre-determined datatypes.
+template <typename DataTypeT, int NumElems> struct test {
+  void operator()(const Logger &log, sycl::queue &queue, bool &passed) {
```
```suggestion
  void operator()(const Logger &log, sycl::queue &queue, bool &passed) const SYCL_ESIMD_KERNEL {
```

vasilytric replied:
I've added `SYCL_ESIMD_KERNEL` to lambda functions in this commit: 35fd849b6fc222cb2da9d90a1dd6185de07651e7

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_default.cpp:
```diff
@@ -0,0 +1,171 @@
+
+int main(int argc, char **argv) {
+  logger::Logger logger{};
+  sycl::queue queue{};
```
```suggestion
  sycl::queue queue{esimd_test::ESIMDSelector{}, esimd_test::createExceptionHandler()};
```
Would you mind to check if there is anything else we can re-use from the existing generic code?

vasilytric replied:
I've applied this comment in this commit: 35fd849b6fc222cb2da9d90a1dd6185de07651e7

vasilytric committed:
[SYCL][ESIMD] Small improvements: add namespaces that correspond to encapsulation levels

vasilytric committed:
[SYCL][ESIMD] Refactoring of "test" function

Add some struct that calls simd constructor, this let easily add new or
delete existing test case

vasilytric committed:
[SYCL][ESIMD] Return passed status from all functions

vasilytric committed:
[SYCL][ESIMD] Add function for obtain chosen types in named_type_pack

vasilytric committed:
[SYCL][ESIMD] Use string representation of current data type in log function

yuriykoch commented on SYCL/ESIMD/api/functional/logger.hpp:
```diff
@@ -0,0 +1,53 @@
+namespace logger {
+
+// Class that provide methods for printing verbose messages to the console
+struct Logger {
```
You already have the 'logger' namespace, so what's the intent of this class? You might as well write something like `log::fail` using the namespace.

vasilytric replied:
this comment was applied in this commit: 3787c4fc6d62817f7c4e1458c2e5255a60beeef4

yuriykoch commented on SYCL/ESIMD/api/functional/type_coverage.hpp:
```diff
@@ -0,0 +1,112 @@
+
+// Constructing named_type_pack with types that depent on chosen field of
+// tested_types enumeration
+template <tested_types T> auto get_tested_types() {
```
```suggestion
template <tested_types required> auto get_tested_types() {
```
To use `T` for enum is a bit misleading

vasilytric replied:
this comment was applied in this commit:  334063fc311a13b4bb841d02ad12da7cd0ab26e3

yuriykoch commented on SYCL/ESIMD/api/functional/type_coverage.hpp:
```diff
@@ -0,0 +1,112 @@
+      : names{std::forward<nameListT>(nameList)...} {}
+};
+
+enum class tested_types { all_types, fp_types, uint_types, sint_types };
```
```suggestion
enum class tested_types { all, fp, uint, sint };
```
Minor, so feel free to ignore if you feel it right.
The reason I would suggest it because we already have `types` stated once, and we are going to use it with the `tested_types::` prefix always. So we could use `tested_types::all` instead of `tested_types::all_types`. I have a feeling we could have a bit more concise code lines by handling namespaces as the part of the the instance name.

vasilytric replied:
this comment was applied in this commit:  334063fc311a13b4bb841d02ad12da7cd0ab26e3

yuriykoch commented on SYCL/ESIMD/api/functional/type_coverage.hpp:
```diff
@@ -0,0 +1,112 @@
+// tested_types enumeration
+template <tested_types T> auto get_tested_types() {
+  if constexpr (T == tested_types::all_types) {
+    auto all_types =
```
```suggestion
    const auto all_types =
```
Here and elsewhere

vasilytric replied:
this comment was applied in this commit:  334063fc311a13b4bb841d02ad12da7cd0ab26e3

yuriykoch commented on SYCL/ESIMD/api/functional/type_coverage.hpp:
```diff
@@ -0,0 +1,112 @@
+        {"float", "sycl::half", "double"});
+    return fp_types;
+  } else if constexpr (T == tested_types::uint_types) {
+    if constexpr (!std::is_signed<char>::value) {
```
```suggestion
    if constexpr (!std::is_signed_v<char>) {
```
We can be a bit more C++17ish, because SYCL-2020 is based on C++17) Here and elsewhere

vasilytric replied:
Yes, good hint, thanks. this comment was applied in this commit:  334063fc311a13b4bb841d02ad12da7cd0ab26e3

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_default.cpp:
```diff
@@ -0,0 +1,167 @@
+  static std::string get_description() { return "initializer"; }
+
+  template <typename DataT, int NumElems>
+  static void call_simd_ctor(DataT *output_data) SYCL_ESIMD_KERNEL {
```
That's interesting. Do we really need the `SYCL_ESIMD_KERNEL` for every function used in kernel? I assumed we need it only for the kernel itself..

vasilytric replied:
this comment was applied in this commit: c201b251096c6d476ae70c763abe09fd4cd801d8

yuriykoch commented on SYCL/ESIMD/api/functional/logger.hpp:
```diff
@@ -0,0 +1,53 @@
+  static void fail(const std::string &msg,
+                   std::string underlying_data_type = "") {
+    std::string log_msg{msg};
+    if (underlying_data_type != "") {
```
```suggestion
    if (!underlying_data_type.empty()) {
```
I have a feeling such API will be a bottle-neck for the following tests, so we might need to provide more generic interface someday..

vasilytric replied:
this comment was applied in this commit: 98d99192e62693a3205e15f51dc5a4bd3656f338

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_default.cpp:
```diff
@@ -0,0 +1,167 @@
+using namespace cl::sycl;
+using namespace sycl::ext::intel::experimental;
+using namespace sycl::ext::intel::experimental::esimd;
+using namespace esimd_test::api::functional::type_coverage;
+using namespace esimd_test::api::functional::logger;
```
```suggestion
```
Do you really need a separate `type_coverage` namespace or a `using namespace logger` expression here?

vasilytric replied:
this comment was applied in this commit: 3787c4fc6d62817f7c4e1458c2e5255a60beeef4

vasilytric committed:
[SYCL][ESIMD] Small refactoring test for simd default constructor

Base includes moved into common.hpp, removed "type_coverage" namespace
and deleted "Logger struct", them methods now just a functions

vasilytric committed:
[SYCL][ESIMD] Small changes in type_coverage.hpp

- Now return named_type_pack without constructing some variable
- change names fields of tested_types enum

vasilytric committed:
[SYCL][ESIMD] Use string.empty() instead of strings comparison

vasilytric committed:
[SYCL][ESIMD] Remove SYCL_ESIMD_KERNEL for functions used in kernel

vasilytric committed:
[SYCL][ESIMD] Remove "using namespace sycl::ext::intel::experimental;"

vasilytric committed:
[SYCL][ESIMD] Apply clang-format

vasilytric commented:
@kbobrovs @v-klochkov could you review this test for `esimd` default constructor?

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_default.cpp:
```diff
@@ -0,0 +1,158 @@
+//
+// Test for esimd default constructor.
+
+#include "common.hpp"
```
Sorry to ask, but I don't see this file here

vasilytric replied:
Sorry, I've miss this file in git includes
I've added this file in this commit: 9890c4c1642c2f97517487c983e9d4fc35da6334

yuriykoch commented on SYCL/ESIMD/api/functional/type_coverage.hpp:
```diff
@@ -0,0 +1,100 @@
+// Type pack to store types and underlying data type names to use with
+// type_name_string
+template <typename... T> struct named_type_pack {
+  const std::string names[sizeof...(T)];
```
Doesn't this file miss the `string` include?

vasilytric replied:
I've applied this comment in this commit: ce387bacfd1ee92e9e59ec2d8f9294b5d77223e3

yuriykoch commented on SYCL/ESIMD/api/functional/type_coverage.hpp:
```diff
@@ -0,0 +1,100 @@
+
+  template <typename... nameListT>
+  named_type_pack(nameListT &&... nameList)
+      : names{std::forward<nameListT>(nameList)...} {}
```
Please include the `utility` C++ header

vasilytric replied:
I've applied this comment in this commit: ce387bacfd1ee92e9e59ec2d8f9294b5d77223e3

yuriykoch commented on SYCL/ESIMD/api/functional/type_coverage.hpp:
```diff
@@ -0,0 +1,100 @@
+    return named_type_pack<float, sycl::half, double>(
+        {"float", "sycl::half", "double"});
+  } else if constexpr (required == tested_types::uint) {
+    if constexpr (!std::is_signed_v<char>) {
```
Please include `type_traits` header

vasilytric replied:
I've applied this comment in this commit: ce387bacfd1ee92e9e59ec2d8f9294b5d77223e3

vasilytric committed:
[SYCL][ESIMD] Move includes from *.cpp file into common.hpp

vasilytric committed:
[SYCL][ESIMD] Update logger.hpp and type_coverage.hpp

Add "#pragma once" and some includes into logger.hpp and type_coverage.hpp

vasilytric committed:
[SYCL][ESIMD] Apply clang-format for type_coverage.hpp

yuriykoch commented:
I would still strongly recommend to use `api/functional/common.hpp` instead of `api/functional/ctors/common.hpp`
as discussed [earlier](https://github.com/intel/llvm-test-suite/pull/565#discussion_r752183369).
I can see no reason why we don't need such header for other functional tests and why we should move it later.

vasilytric committed:
[SYCL][ESIMD] Move common.hpp into  SYCL/ESIMD/api/functional

vasilytric commented:
> I would still strongly recommend to use api/functional/common.hpp instead of api/functional/ctors/common.hpp
as discussed earlier.
I can see no reason why we don't need such header for other functional tests and why we should move it later.

OK, I moved this file into `SYCL/ESIMD/api/functional` in this commit: 18c93a183b2f1580838d3fd0d675ab863cba9482

vasilytric committed:
[SYCL][ESIMD] Update common.hpp

Fix text of the file description, update includes: fix path in includes
due to some path have extra "../" and it is not correspond to current
path, on Windows it was OK, but on Linux it could be compilation error

v-klochkov commented:
Please fix the test, see the CI fail at the newly added test.

v-klochkov commented on SYCL/ESIMD/api/functional/type_coverage.hpp:
```diff
@@ -0,0 +1,106 @@
+//===-- logger.hpp - Declarate functions and containers with that will be used
```
Please copy-paste error here, this line has wrong file name and probably description as well.

vasilytric replied:
The description of `type_coverage.hpp`-file was changed [here](https://github.com/intel/llvm-test-suite/pull/565/commits/d56ee8f4b7fd7c3b8f8d3015a45caef1e6bda23e#diff-3bbc1ab1445c827282cc08d73f76b58cf553d9ced2c799ba46e6148583896d22L1-R10)

v-klochkov replied:
Thank you. I am marking this as resolved now.

v-klochkov commented on SYCL/ESIMD/api/functional/ctors/ctor_default.cpp:
```diff
@@ -0,0 +1,158 @@
+// Dummy kernel for submitting some code into device side.
+template <typename DataT, int NumElems, typename T> struct Kernel;
+
+// Struct that provide method that let get current test case description and
+// method that let call simd constructor for initializer test case
```
provide -> provides
let get current -> lets getting the current
let call -> lets calling
```suggestion
// This struct provides a method returning the test case description and a worker method
// testing the default simd constructor used in RHS of assignment.
```
Please consider fixing descriptions of similar structs  below.

vasilytric replied:
May be mentioning a logic of implementation is not a better Idea... So, comments for these structures were reworked
they can be found [here](https://github.com/intel/llvm-test-suite/pull/565/commits/d56ee8f4b7fd7c3b8f8d3015a45caef1e6bda23e#diff-6d44d6893e30c56cec1c2e698ea43849081d8b092f7a87532bd567294d14ed69L28-L67)

vasilytric committed:
[SYCL][ESIMD] Update comments

- Were reworked comments that can rot
- Were changed some other comments

vasilytric committed:
[SYCL][ESIMD] Remove unnecessary kernel

vasilytric committed:
[SYCL][ESIMD] Add TEST_HALF macros for sycl::half variables

simd can't be constructed with sycl::half data type, i've done it for
successfully running GitHub CI

vasilytric committed:
[SYCL][ESIMD] Use comma separator in list required platforms for testing

List of required platforms should be separated by commas, it is required
by the document "LLVM Testing Infrastructure Guide"

vasilytric committed:
[SYCL][ESIMD] Add comment that simd constructor not support sycl::half

vasilytric commented:
> Please fix the test, see the CI fail at the newly added test.

The CI on `Linux` and `Windows` are passed, CI on `cuda` failed (expected to fail due to failures in previous runnings) due to another tests failures but test for simd constructor should be skipped on `cuda`
For issue with `sycl::half` was created bug on intel/llvm: [intel/llvm/issues/5077](https://github.com/intel/llvm/issues/5077)

Also at the beginning of the tests was added comment about situation with `sycl::half`, it can be found [here](https://github.com/intel/llvm-test-suite/pull/565/commits/d1ab19eaa124415e4f6dc4864151cccbc76f2428#diff-6d44d6893e30c56cec1c2e698ea43849081d8b092f7a87532bd567294d14ed69R14-R17)

@v-klochkov do you have any other comments to this PR? can we merge it?

v-klochkov commented on SYCL/ESIMD/api/functional/ctors/ctor_default.cpp:
```diff
@@ -0,0 +1,156 @@
+//
+// Test for esimd default constructor.
+//
+// The simd constructor can't be constructed with sycl::half data type. The
```
Minor/non-blocking comment:
```suggestion
// The simd can't be constructed with sycl::half data type. The
```
or
```suggestion
// The simd constructor can't be called with sycl::half data type now. The
```

vasilytric replied:
I've applied this comment was applied in this commit: 11cca0a5e06e3913eafa264395c723e135c1636c

v-klochkov commented on SYCL/ESIMD/api/functional/ctors/ctor_default.cpp:
```diff
@@ -0,0 +1,156 @@
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+// REQUIRES: gpu, level_zero
```
```suggestion
// REQUIRES: gpu
```
I don't see anything special in this test that would prevent it from running with opencl. Majority of ESIMD tests can run on both level_zero and opencl, e.g.: https://github.com/intel/llvm-test-suite/blob/intel/SYCL/ESIMD/vadd_1d.cpp#L8

vasilytric replied:
Due to OPenCL is not supported on simd constructor tests we can't use this requirements, but was added these strings:
```
// XREQUIRES: gpu
// TODO gpu and level_zero in REQUIRES due to only this platforms supported yet.
// The current "REQUIRES" should be replaced with "gpu" only as mentioned in
// "XREQUIRES".
```

they can be found in the commit: 84aa45dfae2fabf7818cc522be80a357d061e513

vasilytric committed:
[SYCL][ESIMD] Use comma-separator on unsupported platforms

vasilytric committed:
[SYCL][ESIMD] Update comment about issue with simd constructor

vasilytric committed:
SYCL][ESIMD] Add "XREQUIRES" field and description of unsupp platform

We requiring gpu and level_zero platforms due to the tests on simd
constructors doesn't supported OPenCL platform yet

vasilytric commented:
@v-klochkov  CI on windows failed due to `built-ins/vector_relational.cpp` test, the test `simd_ctor_default.cpp` is not failed, so this failure is not related to reviewed test

vasilytric commented:
@v-klochkov can we merge this PR?
